iothub/device/src/Transport/Amqp/AmqpUnit.cs

```diff
@@ -525,25 +547,31 @@ public async Task EnableEventReceiveAsync(CancellationToken cancellationToken)
 
             try
             {
-                if (_eventReceivingLink == null
-                    || _eventReceivingLink.IsClosing())
+                if (_eventReceivingLink == null || _eventReceivingLink.IsClosing())
                 {
+                    // SafeClose is a fire-and-forget operation. As a result, when it returns the AMQP link might be in closing state
+                    // and may still be referenced by its parent object. Adding locks or checks for this isn't possible because the AMQP
+                    // library doesn't provide any callbacks for notifying us of the state.
+                    // Instead, we have error handling logic when we open links or try to perform operations on opened links.
+                    // If the operation throws an exception, the error handling code will determine if it is to be tried, and it will retry, if necessary.
+                    // This call to SafeClose is necassry because the AMQP library does not call SafeClose immediately after a link closure was requested (as a part of its link lifecycle)
+                    // but instead calls SafeClose eventually. Opening a link with the same name as one previously opened will give rise to ResourceLocked conflicts.
+                    // Another way to avoid ResourceLocked conflicts is to open links with unique names. This approach was previously adopted but later modified in an attempt to reduce link name length.
                     _eventReceivingLink?.SafeClose();
 
-                    _eventReceivingLink = await amqpIotSession
-                        .OpenEventsReceiverLinkAsync(_deviceIdentity, cancellationToken)
-                        .ConfigureAwait(false);
+                    _eventReceivingLink = await _amqpIotSession.OpenEventsReceiverLinkAsync(_deviceIdentity, cancellationToken).ConfigureAwait(false);
 
                     if (_eventReceiverLinkDisconnected == null)
                     {
                         _eventReceiverLinkDisconnected = (obj, arg) =>
                         {
-                            amqpIotSession.SafeClose();
+                            _amqpIotSession.SafeClose();
                         };
                     }
 
-                    _messageReceivingLink.Closed += _eventReceiverLinkDisconnected;
+                    _eventReceivingLink.Closed += _eventReceiverLinkDisconnected;
```

Comment:
This was probably a merge edit at some point. The horton pipeline caught this bug. A win for horton!